### PR TITLE
Batch norm layer for pseudo-fp16 support.

### PR DESCRIPTION
**Pull request by csarofeen**
Batch norm layer that keeps parameters in 32-bit when using 16-bit input/output. This layer is necessary for successful pseudo fp16 training.

### CODE REVIEW

**Review by soumith — APPROVED** [review]
thanks for introducing `nn.contrib`. Long-standing needed change.

**Comment by ngimel** [comment]
It generates unused variable warnings, because now all cudnn routines call ```PyObject* accumTensorClass = getAccumTensorClass(args);``` even though only batch norm uses differently typed accum tensors. Could not figure out how to get around it, would appreciate if someone gave me pointers how to do that in cwrap.

**Comment by csarofeen** [comment]
Removed BN layer in torch/nn left backend in place.

File: .gitmodules [hunk 1]
```diff
@@ -1,6 +1,3 @@
-[submodule "torch/lib/gloo"]
```
Comment (soumith) [inline .gitmodules]: unremove

**Comment by apaszke** [comment]
I'm not sure what's going on in this change. It's changing the BatchNorm code path in C++, but stats will still be half on the Python side? Also, it's going to conflict with @ezyang changes (to no longer use `THVoidTensor`)

**Comment by soumith** [comment]
it's a rebased branch to get Zach going with Volta fp16.
After ed's changes all go in, this can be rebased on top.

**Comment by apaszke** [comment]
Yeah, but this is lacking Python changes. I can't see how will it not error out when you try to use BatchNorm

**Comment by ngimel** [comment]
@apaszke, It does not touch current batch norm in any way, just allows to call cudnn batch norm with fp16 inputs from a custom python module. It will have to be changed after Ed's changes go in.

File: .gitmodules (hunk 1 above)
Comment (csarofeen) [reply .gitmodules]: :-( how did that get in there

**Comment by csarofeen** [comment]
Replaced with https://github.com/pytorch/pytorch/pull/4021 due to cuDNN rewrite in aten.
